hardware_pydantic/devices.py

```diff
@@ -39,8 +72,7 @@ def projection__transfer_between_vials(self, from_obj: Vial, to_obj: Vial, amoun
         return amount / transfer_speed
 
 
-class VialTransferor(Device):
-
+class VialGripperArm(Device):
     def projection__transfer(
```

Comment:
Can you briefly explain the `projection__` prefix and `action__` prefix to me?

Reply:
The life cycle of an `action`:
1. An `instruction` is sent to a `casymda_hardware.DeviceBlock`
2. (to be done) `device.precondition__<action_name>` checks `lab.state`, raise `PreconditionErrror` if not met
3. `device.projection__<action_name>` calculates the time it would need to finish `<action_name>`. A projection method always returns a float
4. (to be done) `device.request__<action_name>` occupies all involved objects
5. `simpy.timeout` lets the time pass in simulation
6. `device.action__<action_name>` changes objects' states, (to be done) raise `ActionError` if transitions are illegal
7. (to be done) `device.release__<action_name>` occupies all involved objects
8.  the `instruction` exists from `DeviceBlock`

Reply:
Note on `occupy` and `release`: currently, only `Device` instances are mapped to `simpy.Resource` instances so there is no way to occupy `LabObject` instances in `simpy`. A `device` is occupied 
1. after it receives an `instruction`
2. the `instruction` is at the top of its queue (first come first serve)

Reply:
Is there a specific reason to separate out `device.request__<action_name>` rather than have that logic be handled within a certain `device.action__<action_name>`? This would also let us release LabObjects part of the way through an action if possible. I'm just wondering about the proliferation of different methods.

To your final point, it seems like we need each `LabObject` to be a `simpy.Resource` then?

Reply:
For occupying LabObjects and devices, does `simpy.Resource` support the use of `with` context managers?

Reply:
> Is there a specific reason to separate out `device.request__<action_name>` rather than have that logic be handled within a certain `device.action__<action_name>`? This would also let us release LabObjects part of the way through an action if possible. I'm just wondering about the proliferation of different methods.

my intention was to separate `simpy` related methods from hardware (state transition) methods... The latter can work without `simpy`.

> To your final point, it seems like we need each `LabObject` to be a `simpy.Resource` then?

this was a point we discussed yesterday: we don't want a robot arm to come and grab the vial that will be used for dispensing. to do so we need to make the vial (a `LabObject`) a `simpy.Resource` so it can be occupied by the dispenser.

Reply:
>does simpy.Resource support the use of with context managers?

the events of `simpy.Resource` (put/get queue) are all context managers already